linbox/algorithms/lifting-container.h

```diff
@@ -152,12 +152,15 @@ namespace LinBox
             Integer N, D, L, Prime;
             this->_intRing.convert(Prime,_p);
 
-            auto hadamardBound = RationalSolveHadamardBound(A, b);
-            N = Givaro::pow(Integer(2), hadamardBound.numBoundBitSize);
-            D = Givaro::pow(Integer(2), hadamardBound.denBoundBitSize);
-			L = N * D * 2;
-
-			_length = (size_t)logp(L,Prime) + 1;   // round up instead of down
+            auto hb = RationalSolveHadamardBound(A, b);
+            N = Givaro::pow(Integer(2), static_cast<size_t>(std::ceil(hb.numLogBound)));
+            D = Givaro::pow(Integer(2), static_cast<size_t>(std::ceil(hb.denLogBound)));
+
+            // L = N * D * 2
+            // _length = logp(L, Prime) = log2(L) * ln(2) / ln(Prime)
+            double primeLog = Givaro::naturallog(Prime);
```

Comment:
why take the natural log if you really want the log2 (line below divides by log(2)). There is a log_2 function in givaro (gmp++)

Reply:
or Givaro::logtwo which you are using later on...

Reply:
Updated.